docs/advanced/signal_generators.md

```diff
@@ -0,0 +1,558 @@
+
+```{note}
+SignalGeneratorMixin only contains abstract methods, defining the class by itself and calling methods in it will only raise NotImplemented errors.
+```
+
+```{mermaid}
+classDiagram
+    direction LR
+
+    Tekscope <|-- InternalAFGChannel
+    AFG <|-- AFGSourceChannel
+
+    AWG <|-- AWGSourceChannel
+    AWGSourceChannel <|-- AWG5KSourceChannel
+    AWG5KSourceChannel <|-- AWG7KSourceChannel
+    AWGSourceChannel <|-- AWG5200SourceChannel
+    AWGSourceChannel <|-- AWG70KSourceChannel
+```
+
+Each Source class (`AFG`, `AWG`) and `Tekscope` (If the AFG license is installed) will contain a dictionary of source channel classes,
+which are defined on first access. Each of these source channel classes represents an output on the source, or the IAFG in the case
+of an oscilloscope.
+
+These channels contain methods and properties which pertain to PI commands which only apply changes to one output channel.
+For example: the afg.source_channel\["SOURCE1"\].set_amplitude() call will change the amplitude only for source output 1.
+
+```{tip}
+These channel classes not only provide easy access to basic scipy commands, but also helper functions, like `set_function_properties`
+```
+
+______________________________________________________________________
+
+### Class Methods
+
+```{warning}
+Each method exclude most attempts at validation, as the end user can change aspects outside its purview.
+There are several distinct instances where this can cause unwanted behavior, depending on the source and what
+state it was in before it is used. Attempting validation when changes can occur outside it's scope leads to many redundant checks.
+As such, it is up to the user to implement these checks.
+```
+
+Each class has children which inherit the base abstracted methods. These methods are tailored to each signal generator so the
+methods handle similarly, regardless of the different PI commands required.
+
+- `source_device_constants` is a property which holds information about what functions
+  and memory sizes are allowed.
+
+```{tip}
+`source_device_constants.functions` will provide an enum of possible functions to generate on the signal generator used.
+```
+
+`generate_function` is a method which allows for the user to request a function from
+any source channel provided an amplitude, frequency and offset is supplied. Other key features
+include the ability to manipulate specific aspects of certain functions. Ramp waveforms can have their symmetry changed
+and duty cycle can be altered for pulse functions. The termination of the IAFG and any AFG can be
+specified using `HIGHZ` or `FIFTY` string literals. If the output needs to be inverted,
+the polarity can be changed on AFGs.
+
+```{warning}
+`generate_function` allows function parameters which can exceed actual generation bounds.
+`get_waveform_constraints` should be used in tandem with `generate_function`, or utilizing the constraints provided in
+[Signal Generators](#signal-generators).
```

Comment:
Should this link to the constraints section?

Reply:
Maybe indicate the constraints are below as well.

Reply:
Unsure what this means